IterateOnEntrySetRule.java

```diff
@@ -0,0 +1,64 @@
+package com.amazonaws.guru.rules.javabestpractices;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class IterateOnEntrySetRule {
+    //should not flag
+    public void IterateOnKeySet_NotUsingValues() {
+        Map<String,String> gfg = new HashMap<String,String>();
+        for (String name : gfg.keySet())
+            System.out.println("key: " + name);
+    }
+    //should flag
+    public void IterateOnKeySet_UsingValues() {
+        Map<String,String> gfg = new HashMap<String,String>();
+        for (String name : gfg.keySet())
+            System.out.println("Value: " + gfg.get(name));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.